src/georaster-layer-for-leaflet.ts

```diff
@@ -0,0 +1,592 @@
+    const heightOfSampleInScreenPixels = this.tileHeight / numberOfSamplesDown;
+    const heightOfSampleInScreenPixelsInt = Math.ceil(heightOfSampleInScreenPixels);
+    const widthOfSampleInScreenPixels = this.tileWidth / numberOfSamplesAcross;
+    const widthOfSampleInScreenPixelsInt = Math.ceil(widthOfSampleInScreenPixels);
+
+    const map = this.getMap();
+    const tileSize = this.getTileSize();
+
+    // this converts tile coordinates (how many tiles down and right)
+    // to pixels from left and top of tile pane
+    const tileNwPoint = coords.scaleBy(tileSize);
+
+    // render asynchronously so tiles show up as they finish instead of all at once (which blocks the UI)
+    setTimeout(async () => {
+      let tileRasters: Georaster[];
+      if (!rasters) {
+        tileRasters = await this.getRasters({
+          tileNwPoint,
+          heightOfSampleInScreenPixels,
+          widthOfSampleInScreenPixels,
+          coords,
+          pixelHeight,
+          pixelWidth,
+          numberOfSamplesAcross,
+          numberOfSamplesDown,
+          ymax,
+          xmin
+        });
+      }
+
+      for (let h = 0; h < numberOfSamplesDown; h++) {
+        const yCenterInMapPixels = tileNwPoint.y + (h + 0.5) * heightOfSampleInScreenPixels;
+        const latWestPoint = L.point(tileNwPoint.x, yCenterInMapPixels);
+        const { lat } = map.unproject(latWestPoint, coords.z);
+        if (lat > yMinOfLayer && lat < yMaxOfLayer) {
+          const yInTilePixels = Math.round(h * heightOfSampleInScreenPixels);
+
+          let yInRasterPixels: number | null;
+          if (inSimpleCRS || this.projection === EPSG4326) {
+            yInRasterPixels = Math.floor((yMaxOfLayer - lat) / pixelHeight);
+          } else {
+            yInRasterPixels = null;
+          }
+
+          for (let w = 0; w < numberOfSamplesAcross; w++) {
+            const latLngPoint = L.point(tileNwPoint.x + (w + 0.5) * widthOfSampleInScreenPixels, yCenterInMapPixels);
+            const { lng: xOfLayer } = map.unproject(latLngPoint, coords.z);
+            if (xOfLayer > xMinOfLayer && xOfLayer < xMaxOfLayer) {
+              let xInRasterPixels: number;
+              if (inSimpleCRS || this.projection === EPSG4326) {
+                xInRasterPixels = Math.floor((xOfLayer - xMinOfLayer) / pixelWidth);
+              } else if (this.getProjector()) {
+                const inverted = this.getProjector().inverse({ x: xOfLayer, y: lat });
+                const yInSrc = inverted.y;
+                yInRasterPixels = Math.floor((ymax - yInSrc) / pixelHeight);
+                if (yInRasterPixels < 0 || yInRasterPixels >= rasterHeight) continue;
+
+                const xInSrc = inverted.x;
+                xInRasterPixels = Math.floor((xInSrc - xmin) / pixelWidth);
+                if (xInRasterPixels < 0 || xInRasterPixels >= rasterWidth) continue;
+              }
+
+              let values = null;
+              if (tileRasters) {
```

Comment:
The compiler is complaining that `tileRasters` is used before being defined. Underwhat conditions is `tileRaster` undefined and what should happen when it is? 

Could we declare a single variable `tileRasters` and assign it either to `this.rasters` or `this.getRasters` depending on some condition, then use that variable in the `for` block instead of checking what exists again on line 324-332? That way it will always be defined.

Reply:
tileRasters is undefined if the georaster is loading from a file that is loaded from an in-memory file.  tileRasters will be defined when we are using a georaster loaded from a url and tileRasters will refer to just the parts of the raster that are fetched to fulfill the tile request.

Reply:
@Rennzie , could we default tileRasters to `null`?  Would that make the compiler happy?

Reply:
Yeah nice one. Defaulting to `null` solved it.